File: DOTNET_CSharp/Library/HotelBookingsPossible.cs
using System.Linq;
using static Library.EventType;

namespace Library {
    public enum EventType { Arrival, Departure }
    public static class HotelBookingsPossible {
        public static bool Check(
            int[] arrivals,
            int[] departures,
            int totalRooms)
        {
            var arr = arrivals.Select(a => (date: a, type: Arrival));
            var dep = departures.Select(d => (date: d, type: Departure));
            var events = arr.Union(dep)
                            .OrderBy(e => e.date);

            var count = 0;
            foreach (var (date, type) in events)
            {
                if (type == Arrival) count++;
                else count--;

                if (count > totalRooms) return false;
            }
            return true;
        }
    }
}

File: FSharp/Library/HotelBookingsPossible.fs
namespace Library

module HotelBookingsPossible =
    type EventType = Arrival | Departure
    type Event = { Date: int; Type: EventType }

    let Check arrivals departures totalRooms =
        let arr = 
            arrivals 
            |> List.map (fun a -> { Date = a; Type = Arrival })
        
        let dep = 
            departures 
            |> List.map (fun d -> { Date = d; Type = Departure })

        let events = 
            arr @ dep 
            |> List.sortBy (fun e -> e.Date)
        
        let mutable count = 0
        let mutable flag = true

        for e in events do
            if e.Type = Arrival then 
                count <- count + 1
            else 
                count <- count - 1;
            
            if count > totalRooms then 
                flag <- false
        
        flag
    